### DISCUSSION TITLE
Detect AR availability

### DISCUSSION THREAD

**Discussion by giuliac1989** [google/model-viewer#2377] [Q&A]
Is there a way to detect if AR is available on a device? I tried with `modelViewer.canActivateAR` but it return always `false`.

Thank you!

**Comment by ghost** [comment]
The canActivateAR() should return this, but you do need to indicate which AR backends you allow with `ar-modes="webxr scene-viewer quick-look"`
As in the example: https://modelviewer.dev/examples/augmentedreality/#ar

Beyond that it would depend on the device your using, also if you can provide a glitch example of your setup that would be helpful, thanks!

**Reply by giuliac1989** [reply]
The method `canActivateAR()` get me an error (doesn't exist), but the property `canActivateAR` not.
Below a screenshot of my model. Clicking on "AR" I want to detect if the AR is available in the device, if so I run `modelViewer.activateAR()`, otherwise I do something else. 

![example](https://user-images.githubusercontent.com/33481534/118087071-43a8d780-b3c5-11eb-9919-22418164fa0b.png)

My configuration is
`<model-viewer id="modelViewer" src="../ar-models/scene.glb"
                                camera-controls ar ar-modes="webxr quick-look"
                                shadow-intensity="0" shadow-softness="0" min-camera-orbit="auto auto auto" max-camera-orbit="auto auto auto"
                                camera-target="0m 0m 0.01m" field-of-view="0" min-field-of-view="7deg" max-field-of-view="100deg" exposure="1"></model-viewer>`

**Reply by elalish** [reply]
Yes, that's right, it's a getter method, so you call it like a property. Is that working for you?

**Reply by giuliac1989** [reply]
No, it's not working for me. It seems that it always return false, but I can't figure why. I tried on a device which support AR and  it works with the examples in the model-viewer website.

**Reply by elalish** [reply]
Can you post a link so I can take a look?

**Reply by giuliac1989** [reply]
I realized that I made a mistake! I miss the "scene viewer" mode in the html, so it not works on Android device. If I put `ar-modes="webxr scene-viewer quick-look"` instead of `ar-modes="webxr quick-look"` it works. Thank you!

**Comment by veronicamontezuma** [comment]
hi @elalish, sorry to piggyback on someone else's question, but I do think there might be a problem here... canActivateAR is only returning true for scene-viewer for me as well! 
on webxr/quick-look, everything comes back false. I've tried a few user agents/phones that I know for a fact have AR and it's not working. (I am not actually using scene-viewer, only tested to see if anything would return true)

**Reply by elalish** [reply]
Can you be a bit more specific? A repro link that demonstrates the problem would be very helpful.

**Reply by veronicamontezuma** [reply]
sorry, I'm gonna try to make a glitch to paste here

**Reply by david-rhodes** [reply]
I'm glad I saw this discussion as I was about to open one asking when it's safe to ask `canActivateAR` from model-viewer. 

**Note**: this may or may not be relevant for the discussion here.

If I ask canActivateAR at any time other than after the window being loaded, it returns false (I assume this is the default value). So, if you listen to a window loaded event and then ask, it works as expected, but if you ask immediately after model viewer has been upgraded, it is not working as expected. **Additionally**, and this looks like a bug: if you use `webxr` as priority before `scene-viewer` for `ar-modes` the result will **always** be false!

```html
<script>
  // Too soon, for some reason?
  window.customElements.whenDefined('model-viewer').then(() => {
    var modelViewer = document.querySelector('model-viewer');
    console.log("model-viewer defined: " + modelViewer.canActivateAR);
  });

  // Returns correct result IF webxr is not used first.
  window.addEventListener("load", () => {
    var modelViewer = document.querySelector('model-viewer');
    console.log("window loaded: " + modelViewer.canActivateAR);
  });
</script>
```

```html
// Does not work, EVER.
<body>
  <div class='models-container'>
    <model-viewer style="height: 100%; width: 100%" id="qa-test" ar ar-scale="fixed" camera-orbit="52.5deg 67.7deg 95%"
      ar-modes="webxr scene-viewer quick-look" camera-controls environment-image="neutral"
      src="https://modelviewer.dev/assets/ShopifyModels/Chair.glb" loading="eager">
    </model-viewer>
  </div>
</body>

// Does work, after window is loaded. Why do we need to wait for this?
<body>
  <div class='models-container'>
    <model-viewer style="height: 100%; width: 100%" id="qa-test" ar ar-scale="fixed" camera-orbit="52.5deg 67.7deg 95%"
      ar-modes="scene-viewer webxr quick-look" camera-controls environment-image="neutral"
      src="https://modelviewer.dev/assets/ShopifyModels/Chair.glb" loading="eager">
    </model-viewer>
  </div>
</body>
```

**Reply by elalish** [reply]
What happens when you try to enter AR with WebXR mode? We have some better console logging around this now. Have you met the WebXR requirements? https://modelviewer.dev/examples/augmentedreality/#ar

**Reply by david-rhodes** [reply]
WebXR definitely works on the device I am testing with. 
Screenshot from the console:
![image](https://user-images.githubusercontent.com/23202691/119037132-31a3d780-b977-11eb-8449-d034c2319cf2.png)

What is especially strange about this is that I assume you use this property for showing/hiding the `ar-button`? When in the model-viewer life cycle does that button visibility get toggled based on device capabilities?

**Reply by elalish** [reply]
I think this is a lit-element lifecycle issue. I know there are some subtleties, since everything in async. It seems to be working fine with webXR on our example pages when I check `canActivateAR` in the console. Do you have a link that reproduces the problem?

**Reply by david-rhodes** [reply]
@elalish I've published some examples for you.

https://sftp.viewinyourspace.com/canActivateAR/sceneviewer.html
![image](https://user-images.githubusercontent.com/23202691/119555138-211ea300-bd63-11eb-8d28-c7e58fd4c2b3.png)

https://sftp.viewinyourspace.com/canActivateAR/webxr.html
![image](https://user-images.githubusercontent.com/23202691/119555019-fa606c80-bd62-11eb-97a1-d57c31b1187f.png)

You can check the console with remote debugging enabled. The **only** difference between these two is the order of WebXR in the ar-modes attribute. I'm testing on a Pixel3a XL with Chrome. You'll notice that the button is showing up on both pages (as expected).

If it makes any difference, I have also seen cases where scene-viewer is used in lieu of WebXR, even though WebXR is supported on my device and set as first priority. It's inconsistent, however. Could it be related to this lifecycle issue?

To reiterate, the main thing I'd like to know is: when is it safe to ask `canActivateAR`? 

Edit: as a followup, would it make sense for model-viewer to fire an event when it has determined AR compatibility, if it cannot be determined immediately?

**Reply by elalish** [reply]
Looking at the code, I think I see the problem. We're using WebXR's `isSessionSupported` function, which is async. The trouble is, once that completes then `canActivateAR` is updated, but you don't know when that happens. I'm not sure of the right solution yet. I suppose an event is one option. Thoughts?

**Reply by david-rhodes** [reply]
I'm happy to subscribe to an event. Ideally we'd know right away (it doesn't seem like checking device compatibility would be a computationally expensive process, but I doubt we have control over the async `isSessionSupported` from WebXR). If an event is used, it would be nice to also know (from the event details) which ar-mode has been chosen by model-viewer.

This begs another question, though. Is there a chance that WebXR function does not return before the user has a chance to press the AR button? I ask because I'm definitely seeing cases where model-viewer uses scene-viewer instead of WebXR, but it's non-deterministic.

**Reply by elalish** [reply]
> This begs another question, though. Is there a chance that WebXR function does not return before the user has a chance to press the AR button? I ask because I'm definitely seeing cases where model-viewer uses scene-viewer instead of WebXR, but it's non-deterministic.

It wouldn't be possible using our button (or your own slotted button). However, I'm guessing you're doing something else since you are asking about `canActivateAR`. May I ask why? I think you'll have a simpler and more robust solution using our built-in button system.

**Reply by david-rhodes** [reply]
No, I'm using the slotted button. The `canActivateAR` request is for other purposes. For example, if the device does not support AR, we may want to show a button that presents a QR code modal. 

Here are some repro steps you could try:

1. Set `loading="lazy" reveal="interaction" ar-modes="webxr scene-viewer quick-look"` attributes
2. Add a button as a child of model-viewer element
```html 
<button slot="ar-button"> </button>
```
3. Press the AR button before clicking the poster (so you are not loading the model in the 3d viewer before activating AR).

It almost seems like when this fails that it launches WebXR **and** scene-viewer, one after the other. Is that possible?

**Reply by elalish** [reply]
I can't repro anything like what you're describing on my pixel3 with Chrome. It's not possible for it to launch both WebXR and SceneViewer; it can try one, fail, then try the other. WebXR can fail if its permissions have been denied for the page. If you can get a more consistent repro, please post it here.

**Reply by elalish** [reply]
Also, check the console logs to see what AR modes are trying to activate. We have better messages now in the newest version.

**Reply by david-rhodes** [reply]
Here you go. The first time I opened an asset, it worked as expected in WebXR. The subsequent attempt to open it failed. Maybe the discerning factor here is that I have a gallery of model-viewers on a single page?

Edit: I usually also test in incognito mode to prevent asset caching issues while developing.

![image](https://user-images.githubusercontent.com/23202691/119715084-36f69b80-be29-11eb-9e76-7b5ef2f1637f.png)

**Reply by david-rhodes** [reply]
Here's another log. This time, the first four attempts failed with WebXR, but succeeded with scene-viewer. The 5th attempt couldn't even succeed with scene-viewer and simply hid the ar-button completely. The sixth attempt successfully used WebXR. 

[ar-modes.log](https://github.com/google/model-viewer/files/6549084/ar-modes.log)

**Reply by elalish** [reply]
@klausw Do you have any insights here on the WebXR side?

**Reply by david-rhodes** [reply]
Aside from some obvious lifecycle unpredictability, I'm wondering if it makes sense for `canActivateAR` to also have fallback methods? For example, it currently looks like it reports the value based on the currently selected AR mode. If you ask in WebXR mode and it is either buggy or not supported, it returns false. But, shouldn't it then ask the next ar-mode if it supports AR or not? This way as a developer we have a handle to know if AR is supported by at least one mode. Ideally this happens or is available from some event (`ARCapabilityReported`, perhaps?). The event details can contain information about which modes are supported? Is this feasible? 

The main issue right now is that I'd rather not implement my own AR capability checker that needs to be updated in line with your in the future as that will inevitably lead to bugs. But, in the current form `canActivateAR` is simply not working reliably because of various lifecycle oddities and fallback routines.

Thank you for any assistance or recommendations here!

**Reply by elalish** [reply]
PRs welcome! It'll be easier to figure out the best solution if we have code on a branch of model-viewer to test and comment on.

**Reply by rawwerks** [reply]
@david-rhodes @elalish  - is there a better way to do this? the idea is that if the browser cannot activate AR, display a modal with a QR code and a warning. i'm not having much luck with this working consistently...

```
	window.addEventListener("load", () => {
		var modelViewer = document.querySelector('model-viewer');
		console.log("model-viewer can activate AR: " + modelViewer.canActivateAR);
		if( !Boolean( modelViewer.canActivateAR ) ) {
			modal.style.display = "block";
		} else {
			modal.style.display = "none";
		}
	});
```

**Reply by rawwerks** [reply]
related - do we need to wait for modelViewer.loaded = true before modelViewer.canActivateAR will return the correct value? 

if so, could someone show me the best way to wait for modelViewer to load? unfortunately loaded is a boolean, not an event. right now i'm doing a timeout for 5 seconds after page-load, which seems like a silly workaround.

**Reply by elalish** [reply]
You are already awaiting the load event, so that's happening and AR is not dependent on the model, just the HTML attributes. As for consistency, `canActivateAR` is sadly imperfect since there isn't a way to completely test the AR modes due to privacy. Can you share some examples that are incorrect?

**Comment by ConfidantCommunications** [comment]
The "load" event needs a listener on the model viewer, not the window.
```
<script type="text/javascript">
      var modelViewer = document.querySelector('#model-viewer');
      modelViewer.addEventListener("load", () => {
        console.log("model-viewer can activate AR: " + modelViewer.canActivateAR);
        if(modelViewer.canActivateAR) document.getElementById("qr").classList.add("hidden");
      });

    </script>
```

**Reply by Mrbside** [reply]
this works

> The "load" event needs a listener on the model viewer, not the window.
> 
> ```
> <script type="text/javascript">
>       var modelViewer = document.querySelector('#model-viewer');
>       modelViewer.addEventListener("load", () => {
>         console.log("model-viewer can activate AR: " + modelViewer.canActivateAR);
>         if(modelViewer.canActivateAR) document.getElementById("qr").classList.add("hidden");
>       });
> 
>     </script>
> ```
